File: Algorithms/cpp/quick_sort.cpp
/**
========================================================================
@file quick_sort.cpp

Big O:

    Space Complexity: O(1)
        - no new items/copies created.

    Time Complexity: O(n^2)
        - average case (great performance): O(n log n)
        - worst case (if data is sorted or almost sorted): O(n^2)


: zach wolpe
: zach.wolpe@medibio.com.au
: 17 July 2023
========================================================================
*/


#include <iostream>

void swap(int array[], int firstIndex, int secondIndex) {
    int temp = array[firstIndex];
    array[firstIndex]   = array[secondIndex];
    array[secondIndex]  = temp;
};

int pivot(int array[], int pivotIndex, int endIndex) {
    // loop
    int swapIndex = pivotIndex;
    for (int i = pivotIndex + 1; i <= endIndex; i++) {
        if (array[i] < array[pivotIndex]) {
            swapIndex ++;
            swap(array, swapIndex, i);
        }
    };
    swap(array, pivotIndex, swapIndex);
    return swapIndex;
};

void quickSort(int array[], int leftIndex, int rightIndex) {
    if (leftIndex >= rightIndex) return;
    int pivotIndex = pivot(array, leftIndex, rightIndex);
    quickSort(array, leftIndex, pivotIndex-1);
    quickSort(array, pivotIndex+1, rightIndex);
};

int main() {
    int myArray[] = {4,6,1,7,3,2,5};
    int size = sizeof(myArray) / sizeof(myArray[0]);
    quickSort(myArray, 0, size-1);
    for (auto value : myArray) {
        std::cout << value << " ";
    }
    return 0;
};

File: Algorithms/python/quick_sort.py

"""
========================================================================
@file quick_sort.cpp

Big O:

    Space Complexity: O(1)
        - no new items/copies created.

    Time Complexity: O(n^2)
        - average case (great performance): O(n log n)
        - worst case (if data is sorted or almost sorted): O(n^2)


: zach wolpe
: zach.wolpe@medibio.com.au
: 18 July 2023
========================================================================
"""

class quickSort:

    @staticmethod
    def swap(array, i, j):
        temp = array[i]
        array[i] = array[j]
        array[j] = temp

    @staticmethod
    def pivot(array, pIndex, endIndex):
        swapIndex = pIndex
        for i in range(pIndex+1, endIndex+1):
            if array[i] < array[pIndex]:
                swapIndex += 1
                quickSort.swap(array, swapIndex, i)
        quickSort.swap(array, pIndex, swapIndex)
        return swapIndex
    
    @staticmethod
    def quick_sorter(array, left, right):
        if left < right:
            pivotIndex = quickSort.pivot(array, left, right)
            quickSort.quick_sorter(array, left, pivotIndex-1)
            quickSort.quick_sorter(array, pivotIndex+1, right)
        return array

    @staticmethod
    def quick_sort(array):
        return quickSort.quick_sorter(array, 0, len(array)-1)

if __name__ == '__main__':
    
    array = [1, 222, 233, 4, 5, 6, 97, 8, 9]
    array = quickSort.quick_sort(array)
    [print(i, end=' -> ') for i in array]
    print()